Refactor data generator

mihaimorariu opened the pull request:
This PR refactors the data generator code. The data generator now outputs ground truth boxes with classes and metadata (shape and scale of rescaled image).

0x00b1 commented:
Hi, @mihaimorariu! I appreciate the contribution! I'll review either this evening or tomorrow. Meanwhile, would you mind adding yourself to AUTHORS.rst.

0x00b1 commented:
@mihaimorariu Would you mind rebasing?

mihaimorariu commented:
Hi, @0x00b1! Thank you!

I am new to this, but I would gladly like to contribute as well. I have rebased and added myself to AUTHORS.rst, as requested.

0x00b1 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -1,118 +1,84 @@
+import time
 
-            current_index = (self.batch_index * batch_size) % n
+def scale_shape(shape, min_size=224, max_size=224):
```
Nice! We really needed this! 👍 

You should add a test for this function to `tests.preprocessing.object_detection`.

0x00b1 replied:
It’s a little wonky and unexpected for `shape` to include the channels dimension. I would expect this to work like shape parameters elsewhere (e.g. Keras’ `ImageGenerator`) where it’s a `(width, height)` pair.

0x00b1 replied:
It also has the benefit of not needing separate “channels first” and “channels last” branches.

mihaimorariu replied:
You're right. I have added the unit test.

mihaimorariu replied:
Sure, I can change that. But I think it would be useful to agree on whether we are going to use (width, height) tuples, from now on, or (height, width). I saw different usage in different parts of the code.

0x00b1 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -1,118 +1,84 @@
-                self.batch_index += 1
-            else:
-                current_batch_size = n - current_index
+    return (int(shape[0] * scale), int(shape[1] * scale), shape[2]), scale
```
An implementation that might be clearer:

```Python
    rr, cc = shape

    rr *= scale
    cc *= scale

    return (int(rr), int(cc)), scale
```

0x00b1 replied:
If you drop the channels dimension, you’ll need to drop the channels dimension here too.

mihaimorariu replied:
You're right, this looks cleaner. Fixed.

0x00b1 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -1,118 +1,84 @@
+    target_shape, scale = scale_shape(image.shape, min_size, max_size)
 
-            yield index_array[current_index:current_index + current_batch_size], current_index, current_batch_size
+    return skimage.transform.rescale(image, scale=scale, mode="reflect"), scale
```
I don’t think this function is needed. I’d inline `skimage.transform.rescale`.

0x00b1 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -1,118 +1,84 @@
-    def __next__(self, *args, **kwargs):
-        return self.next(*args, **kwargs)
+class DictionaryIterator(keras.preprocessing.image.Iterator):
+    def __init__(self, data, classes, image_data_generator, image_shape, batch_size=1,
```
I’d keep the keyword “dictionary” as it’s more descriptive than `data`.

0x00b1 replied:
I’d also keep the keyword `generator` since there’re a number of possible generators that could return a `DictionaryGenerator`.

0x00b1 replied:
You don’t need to include the `image_shape` parameter since many datasets have variable shapes (e.g. PASCAL VOC) and this information is included in the dictionary.

mihaimorariu replied:
Agreed. I changed that.

0x00b1 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -1,118 +1,84 @@
-        return self.next(*args, **kwargs)
+class DictionaryIterator(keras.preprocessing.image.Iterator):
+    def __init__(self, data, classes, image_data_generator, image_shape, batch_size=1,
+                 shuffle=True, seed=numpy.uint32(time.time() * 1000)):
```
The inner generators should match the semantics and defaults of other Keras generators (like `DirectoryIterator` or `NumPyIterator`) so `shuffle=False`.

0x00b1 replied:
Likewise, the `seed` semantics should match so it should be `seed=None`.

0x00b1 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -1,118 +1,84 @@
+        self.data                 = data
+        self.classes              = classes
+        self.image_data_generator = image_data_generator
+        self.image_shape          = image_shape
```
`self.image_shape` is unused.

0x00b1 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -1,118 +1,84 @@
+        self.classes              = classes
+        self.image_data_generator = image_data_generator
+        self.image_shape          = image_shape
+        self.target_shape, _      = scale_shape(image_shape)
```
This shouldn’t be inside the constructor since the `target_shape` is dependent on the image.

0x00b1 replied:
You’re missing the call to `super()`.

0x00b1 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -1,118 +1,84 @@
+            # Create metadata
+            metadata[batch_index, :] = [image.shape[0], image.shape[1], scale]
 
+        return [image_batch, gt_boxes_batch, metadata], None
```
You’re missing the `y` value: `[boxes, labels]`

0x00b1 replied:
You’re also missing `labels` from `x`.

0x00b1 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -1,118 +1,84 @@
-        image = skimage.io.imread(pathname)
+        for batch_index, image_index in enumerate(selection):
+            path         = self.data[image_index]["filename"]
+            image        = skimage.io.imread(path, as_grey=(self.target_shape[2] == 1))
```
You don’t need to specify `as_gray`.

0x00b1 commented:
Thanks, @mihaimorariu! This was great. I requested a few changes. Let me know if you need help!

0x00b1 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -1,118 +1,84 @@
-    def flow(self, dictionary, shuffle=True, seed=None):
-        return DictionaryIterator(dictionary, self, shuffle=shuffle, seed=seed)
+    def flow(self, data, classes, image_shape):
+        return DictionaryIterator(data, classes, keras.preprocessing.image.ImageDataGenerator(), image_shape)
```
`keras.preprocessing.image.ImageDataGenerator()` should be `self`

0x00b1 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -1,118 +1,84 @@
-
-    def flow(self, dictionary, shuffle=True, seed=None):
-        return DictionaryIterator(dictionary, self, shuffle=shuffle, seed=seed)
+    def flow(self, data, classes, image_shape):
```
I recommend that we keep using `dictionary` rather than `data`. Is `flow_from_dictionary` is clearer?

0x00b1 replied:
We should drop the `input_shape` parameter.

`shape` should be inferred for situations where images have variable dimensions.

0x00b1 replied:
@hgaiser When shape inference works it’d be `(None, None, 3)`. However, until it’s implemented, the model expects fixed dimensions so you’d need to crop or resize.

0x00b1 replied:
@hgaiser @mihaimorariu Maybe I misinterpreted the `image_size` parameter? Should it be `target_size`?

mihaimorariu replied:
`image_shape` is the shape of the original image. `target_shape` is the shape of its rescaled version. So, if I understand correctly, shall we assume that `image_shape` is the same for all images, for now?

hgaiser replied:
@0x00b1 I get that, but generators work with numpy arrays and numpy arrays don't work with shapes like (None, None, 3). A numpy array has to be created, containing any number of images with any kind of shape, how?

If I remember correctly `image_size` is the original image size and this generator would force all images to be of that size (some check for this still needs to be added, would give a clearer error message too; alternative is to reshape all images to this size). In an ideal situation, `image_size` is not forced to one constant size. However, given the problem mentioned above, the most viable solutions is to either force batch size to 1 or force all images to be of the same size. A combination of the two is something I don't see a solution for. I'm very much open to suggestions on this part 👍

0x00b1 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -1,118 +1,84 @@
+            path         = self.data[image_index]["filename"]
+            image        = skimage.io.imread(path, as_grey=(self.target_shape[2] == 1))
+            image, scale = scale_image(image)
+            image        = self.image_data_generator.random_transform(image)
```
I would drop the `random_transform` and `standardize` calls for the moment.

Unfortunately, we can’t easily inherit from `ImageDataGenerator` since we’ll need to write augmentation and pre-processing functions that can manage images and bounding boxes together. We’ll also need to support masks in the near future.

hgaiser replied:
Just a note, supporting masks is no issue using the `seed` parameter in `random_transform`. The easiest solution for boxes would be to make a fake image where the bounding box is a square segmentation, augment it using `random_transform` and re-compute the bounding box. Not efficient, but easy.

0x00b1 replied:
@hgaiser 👍

0x00b1 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -1,118 +1,84 @@
+                    raise Exception("Class {} not found in '{}'.".format(b["class"], self.classes))
 
-        boxes = numpy.asarray([[d[k] for k in ['y1', 'x1', 'y2', 'x2']] for d in ds])
+                gt_data = [b["y1"], b["x1"], b["y2"], b["x2"], self.classes[b["class"]]]
```
I believe this should be `['y1', 'x1', 'y2', 'x2']` (i.e. without the class label).

0x00b1 replied:
We should be yielding (my current understanding):

```Python
# image: (batches, width, height, channels)
# metadata: (batches, 3) where `3` corresponds to the values of `width`, `height`, and `scale`
# bounding boxes: (boxes, 4) where `4` corresponds to `y1`, `x1`, `y2`, and `x2`
# labels: (boxes, classes)

yield (image, metadata, bounding boxes, labels), (bounding boxes, labels)
```

0x00b1 replied:
(@jhung0 Is my understanding correct?)

jhung0 replied:
yes, boxes should just have the 4 coordinates. labels has the class info

hgaiser replied:
Ah so we're deviating from py-faster-rcnn here? There, `gt_boxes` is a matrix containing the boxes and labels as a (?, 5) matrix.

jhung0 replied:
Yes, the labels here are in one-hot form.

hgaiser replied:
It makes sense to split them, its clearer. Why one-hot form if I may ask?

0x00b1 replied:
@hgaiser It’s the standard for Keras. It does make some stuff easier (a notable exception is ImageNet’s 21,841 classes 😧).

hgaiser replied:
Alright then :)

Figured with the sparse versions of the loss functions, for example `sparse_categorical_crossentropy`, it didn't matter much.

0x00b1 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -1,118 +1,84 @@
 
-    def __next__(self, *args, **kwargs):
-        return self.next(*args, **kwargs)
+class DictionaryIterator(keras.preprocessing.image.Iterator):
```
Nice. 👍 

I think inheriting from ` keras.preprocessing.image.Iterator` will be easier to maintain than continuing to rely on our implementation.

jhung0 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -1,118 +1,87 @@
 
-        pathname = self.dictionary[index]["filename"]
+        # Labels has num_classes + 1 elements, since 0 is reserved for background.
+        num_classes = len(self.classes.keys())
```
could just do len(self.classes)

jhung0 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -5,80 +5,83 @@
+            boxes[batch_index, :, :4] *= self.scale
 
-        return [image_batch, gt_boxes_batch, metadata], None
+        return [images, boxes, labels, self.scale], None
```
Why have a None?

hgaiser replied:
Generators must return `X, Y`, but `Y` doesn't make sense in our case because the network generates `Y`. Therefore, `Y` can be ignored.

jhung0 replied:
Also, shouldn't metadata have image width/height information too?

mihaimorariu replied:
Do we need to include it, since that information can be obtained from `images`?

jhung0 commented on keras_rcnn/preprocessing/_object_detection.py:
```diff
@@ -37,9 +37,14 @@ def __init__(self, dictionary, classes, generator, batch_size=1, shuffle=False,
         cols, rows, channels          = dictionary[0]["shape"]
         self.image_shape              = (rows, cols, channels)
+
         self.target_shape, self.scale = scale_size(self.image_shape[0:2])
```
How would the min/max size for scale_size be set?
